app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,185 @@
+) : Middleware<HomeFragmentState, HomeFragmentAction> {
+
+    private var blocklist: Set<String> = setOf()
+    private var blocklistLoaded = false
+    private var store: Store<HomeFragmentState, HomeFragmentAction>? = null
+
+    init {
+        provider.blocklist.map { updatedList ->
+            blocklist = updatedList
+            if (!blocklistLoaded) {
+                filterLatestState()
+            }
+            blocklistLoaded = true
+        }.launchIn(coroutineScope)
+    }
+
+    /**
+     * Will filter "Change" actions using the blocklist and use "Remove" actions to update
+     * the blocklist.
+     */
+    override fun invoke(
+        context: MiddlewareContext<HomeFragmentState, HomeFragmentAction>,
+        next: (HomeFragmentAction) -> Unit,
+        action: HomeFragmentAction
+    ) {
+        if (!blocklistLoaded) {
+            store = context.store
+            return
+        }
+        next(handleUpdate(context.state, action))
+    }
+
+    private fun handleUpdate(state: HomeFragmentState, action: HomeFragmentAction): HomeFragmentAction = when (action) {
+        is HomeFragmentAction.Change -> {
+            action.copy(
+                topSites = action.topSites.filteredByBlocklist(blocklist),
+                recentBookmarks = action.recentBookmarks.filteredByBlocklist(blocklist),
+                recentTabs = action.recentTabs.filteredByBlocklist(blocklist),
+            )
+        }
+        is HomeFragmentAction.TopSitesChange -> {
+            action.copy(
+                topSites = action.topSites.filteredByBlocklist(blocklist)
+            )
+        }
+        is HomeFragmentAction.RecentTabsChange -> {
+            action.copy(
+                recentTabs = action.recentTabs.filteredByBlocklist(blocklist)
+            )
+        }
+        is HomeFragmentAction.RecentBookmarksChange -> {
+            action.copy(
+                recentBookmarks = action.recentBookmarks.filteredByBlocklist(blocklist)
+            )
+        }
+        is HomeFragmentAction.RecentHistoryChange -> {
+            action.copy(recentHistory = action.recentHistory.filteredByBlocklist(blocklist))
+        }
+        is HomeFragmentAction.RemoveRecentTab -> {
+            if (action.recentTab is RecentTab.Tab) {
+                val url = action.recentTab.state.content.url
+                val updatedBlocklist = blocklist + url
+                provider.addUrlToBlocklist(url)
+                state.filteredByBlocklist(updatedBlocklist).toChangeAction()
+            } else {
+                action
+            }
+        }
+        is HomeFragmentAction.RemoveRecentBookmark -> {
+            action.recentBookmark.url?.let {
+                val updatedBlocklist = blocklist + it
+                provider.addUrlToBlocklist(it)
+                state.filteredByBlocklist(updatedBlocklist).toChangeAction()
+            } ?: action
+        }
+        is HomeFragmentAction.RemoveTopSite -> {
```

Comment:
Reading the jira requirement, it seems like Top Sites should not be filtered here.

Reply:
Removed top sites from the implementation